### PR TITLE
Replace most of our Python lint tools with ruff

### PR DESCRIPTION
**Pull request by ahal**
This depends on the PR to drop support for Python 3.6 because it doesn't have 3.6 as an option in its settings.

### CODE REVIEW

**Review by jcristau — APPROVED** [review]
Cool :)

File: pyproject.toml [hunk 1]
```diff
@@ -7,10 +7,26 @@ extend-exclude = """(\
...
+
+[tool.ruff]
+select = [
+    "E",    # pycodestyle
+    "F",    # pyflakes
+    "I",    # isort
+    "PLC",  # pylint convention
+    "PLE",  # pylint errors
+    "UP",   # pyupgrade
+]
```
Comment (hneiva) [inline pyproject.toml:23]: I suggest adding `"W"` as well
https://beta.ruff.rs/docs/rules/#warning-w

File: src/taskgraph/util/path.py [hunk 2]
```diff
@@ -93,13 +93,8 @@ def basedir(path, bases):
...
 re_cache = {}
-# Python versions < 3.7 return r'\/' for re.escape('/').
-if re.escape("/") == "/":
-    MATCH_STAR_STAR_RE = re.compile(r"(^|/)\\\*\\\*/")
-    MATCH_STAR_STAR_END_RE = re.compile(r"(^|/)\\\*\\\*$")
-else:
-    MATCH_STAR_STAR_RE = re.compile(r"(^|\\\/)\\\*\\\*\\\/")
-    MATCH_STAR_STAR_END_RE = re.compile(r"(^|\\\/)\\\*\\\*$")
+MATCH_STAR_STAR_RE = re.compile(r"(^|/)\\\*\\\*/")
+MATCH_STAR_STAR_END_RE = re.compile(r"(^|/)\\\*\\\*$")
```
Comment (hneiva) [inline src/taskgraph/util/path.py:97]: Depends on #207 ?

**Review by hneiva — APPROVED** [review]

File: pyproject.toml (hunk 1 above)
Comment (hneiva) [reply pyproject.toml:23]: Thinking about this. We should come up with a "releng standard set of rules for ruff"

File: src/taskgraph/util/path.py (hunk 2 above)
Comment (ahal) [reply src/taskgraph/util/path.py]: Yep, sorry I was planning to rebase before requesting review.

File: pyproject.toml (hunk 1 above)
Comment (ahal) [reply pyproject.toml:22]: Good call on adding "W", codebase already passes everything there.

**Pull request merged**